beacon_node/beacon_chain/src/attester_per_slot_service.rs

```diff
@@ -0,0 +1,122 @@
+use crate::{BeaconChain, BeaconChainTypes};
+use slog::{debug, error};
+use slot_clock::SlotClock;
+use std::sync::Arc;
+use task_executor::TaskExecutor;
+use tokio::time::sleep;
+use types::{BeaconState, RelativeEpoch, Slot};
+
+/// Spawns a routine which produces an unaggregated attestation at every slot.
+///
+/// This routine will run once per slot
+pub fn start_attester_per_slot_service<T: BeaconChainTypes>(
+    executor: TaskExecutor,
+    chain: Arc<BeaconChain<T>>,
+) {
+    // TODO: AI(joel) Only run the service if validator monitor is enabled
+    // Paul has made a refacto of that bit in another PR, I will rebase
+    // once it's merged
+    executor.clone().spawn(
+        async move { attester_per_slot_service(executor, chain).await },
+        "attester_per_slot_service",
+    );
+}
+
+/// Loop indefinitely, calling `BeaconChain::produce_unaggregated_attestation` every 4s into each slot.
+async fn attester_per_slot_service<T: BeaconChainTypes>(
+    executor: TaskExecutor,
+    chain: Arc<BeaconChain<T>>,
+) {
+    let slot_duration = chain.slot_clock.slot_duration();
+    let additional_delay = slot_duration / 3;
+
+    loop {
+        match chain.slot_clock.duration_to_next_slot() {
+            Some(duration) => {
+                sleep(duration + additional_delay).await;
+
+                debug!(
+                    chain.log,
+                    "Produce an unaggregated attestation";
+                );
+
+                // I am not sure the impact of cloning the state at every epoch yet
+                // That's probably the naive solution, I'll reiterate on this later
+                let inner_chain = chain.clone();
+                let state = inner_chain.head_beacon_state_cloned();
+
+                // Run the task in the executor
+                executor.spawn(
+                    async move {
+                        if let Ok(current_slot) = inner_chain.slot() {
+                            produce_unaggregated_attestation(inner_chain, state, current_slot);
+                        }
+                    },
+                    "attester_per_slot_service",
+                );
+            }
+            None => {
+                error!(chain.log, "Failed to read slot clock");
+                // If we can't read the slot clock, just wait another slot.
+                sleep(slot_duration).await;
+            }
+        };
+    }
+}
+
+pub fn produce_unaggregated_attestation<T: BeaconChainTypes>(
+    inner_chain: Arc<BeaconChain<T>>,
+    state: BeaconState<<T as BeaconChainTypes>::EthSpec>,
+    current_slot: Slot,
+) {
+    // Get the committee cache for the current slot
+    if let Ok(committee_cache) = state.committee_cache(RelativeEpoch::Current) {
+        let committee_count = committee_cache.committees_per_slot();
+
+        // Produce an unaggregated attestation for each committee
+        for index in 0..committee_count {
```

Comment:
Since attestations for different committees are practically identical (apart from the committee `index` field) we could just produce a single attestation for committee `0`. Committee `0` is guaranteed to exist. That means we could avoid loading the committee cache here, which would be a nice simplification.